docs/extensions.adoc

```diff
@@ -788,25 +789,33 @@ NOTE: It is an error to have multiple configuration objects with the same name,
 probably prefix it with some package-like name to minimize the risk for name clashes with other extensions or the core
 Spock code.
 
-To use the values of the configuration object in your extension, just define an uninitialized instance field of that
-type. Spock will then automatically create exactly one instance of the configuration object per Spock run, apply the
+To use the values of the configuration object in your extension, you can either use constructor injection or field injection.
+
+* To use constructor injection just define a constructor with one or more configuration object parameters.
+* To use field injection just define an uninitialized non-final instance field of that type.
+
+Spock will then automatically create exactly one instance of the configuration object per Spock run, apply the
 settings from the configuration file to it (before the `start()` methods of global extensions are called) and inject
 that instance into the extension class instances.
 
+NOTE: Constructor injection has a higher priority than field injection and Spock won't inject any fields if it finds a suitable constructor.
```

Comment:
Do you see a reason (for this usecase) to support both at the same time?
Normally you don't mix both injections in the same class, at least I've never seen it and didn't have a reason to do so.

> and other objects that are needed later on as fields?

Spock doesn't have multi-stage injection.

I mainly kept field injection for compatibility, but to get rid of those pesky reflective access warnings, you need constructor injection. If we would allow mixed injection, then you couldn't be compatible to both 1.x/2,x, while also getting rid of the reflection warnings.

Reply:
> Normally you don't mix both injections in the same class, at least I've never seen it and didn't have a reason to do so.

I don't agree, I usually do it as I said, if I need something in the constructor, I do constructor injection and otherwise field injection. Also within the same class. This is from my experience a pretty popular tactic when somewhere `@PostConstruct` or similar is not available.

> while also getting rid of the reflection warnings.

Which reflection warnings do you mean exactly and how to reproduce them?
I just played around, made two modules, one (bar) with a class with a public field and constructor with parameter, one (foo) where I instantiate the class from the other module via reflection and set the value of the field and there was no warning.
If I make the field private and try to use `setAccessible` I get an error, but only until I add an `opens bar to foo;` to the module descriptor of bar, then even the setting of the private field works without warning. And it is absolutely fine to require an extension to have an `opens` clause for that purpose and if it is missing it can be given on execution too.

Reply:
And if field injection is only kept for compatibility, maybe it should be made clearer in the docs that it shouldn't be used?

Reply:
Field injection is not removed, you can keep using it for your personal extensions,  I prefer final fields and the certainty that it was properly initialized. 

Furthermore, this is not a general purpose DI framework, we don't have optional dependencies. `@PostConstruct`, ...

You normally have 1 maybe 2 config objects that are injected, so I really don't think that we have to support mixing injection types.